src/common/meta/src/key/topic_region_map.rs

```diff
@@ -0,0 +1,231 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::{self, Display};
+use std::sync::Arc;
+
+use serde::{Deserialize, Serialize};
+use snafu::OptionExt;
+use store_api::storage::RegionId;
+
+use crate::error::{Error, InvalidMetadataSnafu, Result};
+use crate::key::{MetadataKey, TOPIC_REGION_MAP_PATTERN, TOPIC_REGION_MAP_PREFIX};
+use crate::kv_backend::memory::MemoryKvBackend;
+use crate::kv_backend::KvBackendRef;
+use crate::rpc::store::{BatchPutRequest, PutRequest, RangeRequest};
+use crate::rpc::KeyValue;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct TopicRegionMapKey<'a> {
+    pub region_id: RegionId,
+    pub topic: &'a str,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct TopicRegionMapValue;
+
+impl<'a> TopicRegionMapKey<'a> {
+    pub fn new(region_id: RegionId, topic: &'a str) -> Self {
+        Self { region_id, topic }
+    }
+
+    pub fn range_start_key() -> String {
+        TOPIC_REGION_MAP_PREFIX.to_string()
+    }
+}
+
+impl<'a> MetadataKey<'a, TopicRegionMapKey<'a>> for TopicRegionMapKey<'a> {
+    fn to_bytes(&self) -> Vec<u8> {
+        self.to_string().into_bytes()
+    }
+
+    fn from_bytes(bytes: &'a [u8]) -> Result<TopicRegionMapKey<'a>> {
+        let key = std::str::from_utf8(bytes).map_err(|e| {
+            InvalidMetadataSnafu {
+                err_msg: format!(
+                    "RegionTopicMapKey '{}' is not a valid UTF8 string: {e}",
+                    String::from_utf8_lossy(bytes)
+                ),
+            }
+            .build()
+        })?;
+        TopicRegionMapKey::try_from(key)
+    }
+}
+
+impl Display for TopicRegionMapKey<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}/{}/{}",
+            TOPIC_REGION_MAP_PREFIX,
+            self.topic,
+            self.region_id.as_u64()
+        )
+    }
+}
+
+impl<'a> TryFrom<&'a str> for TopicRegionMapKey<'a> {
+    type Error = Error;
+
+    /// Value is of format `{prefix}/{topic}_{region_id}`
+    fn try_from(value: &'a str) -> Result<TopicRegionMapKey<'a>> {
+        let captures = TOPIC_REGION_MAP_PATTERN
+            .captures(value)
+            .context(InvalidMetadataSnafu {
+                err_msg: format!("Invalid region topic map key: {}", value),
+            })?;
+        let topic = captures
+            .get(1)
+            .map(|m| m.as_str())
+            .context(InvalidMetadataSnafu {
+                err_msg: format!("Invalid topic in region topic map key: {}", value),
+            })?;
```

Comment:
It's fine to retrieve the value directly since it has already passed the regex check.
```suggestion
        let topic = captures[1].as_str();
```